keps/sig-storage/20190709-csi-snapshot.md

```diff
@@ -89,28 +92,152 @@ The rest of the document includes required information missing from the original
+  ReadyToUse *bool
+  RestoreSize *int64
+  Error *VolumeSnapshotError
+  SnapshotHandle *string
+}
+```
+
+The beta Kubernetes VolumeSnapshotClass API object is as follows:
+
+```
+type VolumeSnapshotClass struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+
+        Driver string
+        Parameters map[string]string
+        DeletionPolicy DeletionPolicy
+}
+```
+
+### Controller Split
+
+Along with VolumeSnapshot being promoted to Beta in Kubernetes 1.17, the CSI external-snapshotter sidecar controller has been split into two controllers: a snapshot-controller and a CSI external-snapshotter sidecar.
+
+The snapshot controller will be watching the Kubernetes API server for `VolumeSnapshot`, `VolumeSnapshotContent`, and `VolumeSnapshotClass` CRD objects. The CSI `external-snapshotter` sidecar watches the Kubernetes API server for `VolumeSnapshotContent` and `VolumeSnapshotClass` CRD objects.
+
+For dynamic provisioning, the creation of a new `VolumeSnapshot` object referencing a `VolumeSnapshotClass` CRD object corresponding to this driver causes the snapshot controller to trigger the creation of a Kubernetes `VolumeSnapshotContent` object to represent the to-be-created new snapshot.
+
+The creation of a new `VolumeSnapshotContent` object causes the sidecar container to trigger a `CreateSnapshot` operation against the specified CSI endpoint to provision a new snapshot. When a new snapshot is successfully provisioned, the sidecar container updates the status field of the `VolumeSnapshotContent` object to represent the new snapshot.
+
+The snapshot controller will be updating the status field of the `VolumeSnapshot` object accordingly based on the status field of the `VolumeSnapshotContent` object to indicate the new snapshot is ready to be used or failed.
+
+The deletion event of a `VolumeSnapshot` object bound to a `VolumeSnapshotContent` corresponding to this driver with a `delete` deletion policy causes the snapshot controller to start deleting the `VolumeSnapshotContent` object and add an annotation to the object to indicate it is being deleted. Note that both the `VolumeSnapshot` object and the `VolumeSnapshotContent` object will not be deleted immediately due to the finalizers. When the sidecar container detects this update on the `VolumeSnapshotContent` object, it triggers a `DeleteSnapshot` operation against the specified CSI endpoint to delete the snapshot. Once the snapshot is successfully deleted, the sidecar container removes the finalizer on the `VolumeSnapshotContent` object which leads to the deletion of the object from Kubernetes. The snapshot controller then removes the finalizer on the `VolumeSnapshot` object and as a result the object will be deleted from Kubernetes. If a user deletes a bound `VolumeSnapshotContent` object directly, it will have a deletion timestamp set however will persist in API server until its corresponding `VolumeSnapshot` object also gets a deletion timestamp set from a deletion request.
```

Comment:
Why is deletion timestamp not sufficient to determine if the object is being deleted? Why do you also need an annotation?

Reply:
Because the sidecar does not have access to the VolumeSnapshot API object so it can't check if VolumeSnapshot has a deletion timestamp.  The sidecar checks the VolumeSnapshotBeingDeleted annotation to make sure that it is a deletion request on the VolumeSnapshot.

Reply:
How come deletion timestamp on the VolumeSnapshotContent object is not sufficient?

Reply:
We changed the logic in the beta version. 
 
In the Alpha version, we followed PV/PVC logic.  If a user tries to delete VolumeSnapshot, the VolumeSnapshot API object will be deleted immediately if it is not used by a PVC.  For VolumeSnapshotContent, if there's a deletion timestamp and VolumeSnapshot no longer exists, the controller will try to delete the physical snapshot and VolumeSnapshotContent.  However this could potentially delete the physical snapshot by accident when user copies the content yaml to another environment and cause data loss.

In Beta version, VolumeSnapshot will not be deleted before VolumeSnapshotContent and the physical snapshot resource.  We want to make sure that the deletion request is coming from a request to delete the VolumeSnapshot, not directly on VolumeSnapshotContent.  If we are only checking deletion timestamp on VolumeSnapshotContent, then it could leave VolumeSnapshot object behind.

Reply:
Isn't there a finalizer on the Content object that only gets removed when the Snapshot object is deleted? (PV has a finalizer that only gets removed when PVC is gone)